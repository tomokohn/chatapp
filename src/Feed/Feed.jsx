import React from 'react';
import styled from "styled-components";


const FeedContainer = styled.div`
`;

const Feed = ({messages}) => (
  <FeedContainer>
    {messages.map(msg => msg)}
  </FeedContainer>
)

export default Feed;