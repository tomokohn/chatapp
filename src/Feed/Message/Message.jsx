import React from 'react';
import styled from 'styled-components';
import botAvatar from '../../assets/mayas_avatar.png';
import userAvatar from '../../assets/user_avatar.png';


const applyRadius = (last) => last ? '0 15px 15px 15px' : '15px 15px 15px 0';

const MassageContainer = styled.div`
  display: grid;
  grid-template-columns: 50px auto 50px;
  grid-template-rows: minmax(48px, auto);
  align-items: end;
  justify-content: start;
  direction: ${ props => props.bot ? 'ltr' : 'rtl'};
  margin: ${ props => props.bot ? '0px' : '15px 0'};
`;

const Text = styled.p`
  background: ${ props => props.bot ? '#e9e9e9' : '#4a4a4a'};
  padding: 12px;
  border-radius: ${props => applyRadius(props.last)};
  border-radius: ${props => !props.bot && '24px'};
  margin-block-start: 3px;
  margin-block-end: 0;
  grid-column: 2;
  color: ${ props => props.bot ? '#000' : '#fbfbfb'};
`;

const AvatarImg = styled.img`
  width: 40px;
  padding: 0 5px;
`;

const Message = ({text, isBot, showAvatar, last}) => (
  <MassageContainer bot={isBot}>
    { showAvatar ? <AvatarImg src={isBot ?botAvatar : userAvatar} alt=""/> : null}
    <Text last={last} bot={isBot}>{text}</Text>
  </MassageContainer>
);

export default Message;