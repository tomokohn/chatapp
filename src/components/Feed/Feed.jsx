import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from "styled-components";
import { composeMessage } from '../../sevices/messageGenerator.service';
import ScrollBottom from "./ScrollBottom/ScrollBottom";

const FeedContainer = styled.div`
    overflow-y: auto;
    padding-bottom: 15px;
    scroll-behavior: smooth;
    -ms-overflow-style: none;
    &::-webkit-scrollbar {
      display: none;
    }
`;

class Feed extends Component {
  constructor(props) {
    super();
    this.state = {
      showScrollBtn: false
    }
    this.feed = React.createRef();

    this.scrollBottom = this.scrollBottom.bind(this);
    this.handleScroll = this.handleScroll.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    if(prevProps.messages.length !== this.props.messages.length) this.scrollBottom();
  }

  handleScroll() {
    const feed = this.feed.current;
    const shouldshowBtn = feed.scrollHeight > feed.scrollTop + 800;
    if (feed.scrollHeight - feed.scrollTop === feed.clientHeight) {
      this.setState({showScrollBtn: false });
    } else if (shouldshowBtn) this.setState({showScrollBtn: true });
  }

  scrollBottom() {
    const feed = this.feed.current;
    feed.scrollTop = feed.scrollHeight - feed.clientHeight;;
  }

  shouldShowAvatar(isBot, index, array) {
    if (!isBot) return true;
    if (index === (array.length - 1)) return true;
    if (array[index].isBot && array[index + 1].isBot) {
      return false
    } else {
      return true;
    }
  };

  render() {
    return (
      <FeedContainer className="feed" ref={this.feed} onScroll={this.handleScroll}>
        {this.props.messages.map((msg, i, arr) =>
          composeMessage({
            ...msg,
            key: i,
            showAvatar: this.shouldShowAvatar(msg.isBot, i, arr)
          })
        )}
        { this.state.showScrollBtn  && <ScrollBottom onClickDown={()=> this.scrollBottom()}/>}
      </FeedContainer>
    )
  }

}

export default Feed;

Feed.propTypes = {
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string,
      isBot: PropTypes.bool,
      last: PropTypes.bool
    })
  ),
};