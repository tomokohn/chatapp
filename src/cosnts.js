export const MAYA_FIRST_GREETING = 'Hi, Iâ€™m Maya! Today youâ€™re going to help me to ace my game';
export const MAYA_ASKS_NAME = 'Letâ€™s start by telling me your name';
export const MAYA_ASKS_EXPRRSION = 'List any mathematical expression you can think of - Iâ€™ll crunch it in no time';
export const MAYA_ASKS_MORE = 'This was easy, give me something harder ðŸ¤“';

export const greetingMessage = {text: MAYA_FIRST_GREETING, isBot: true};
export const askNameMessage = {text: MAYA_ASKS_NAME, isBot: true, last: true};
export const askExpressionMessage = {text: MAYA_ASKS_EXPRRSION, isBot: true, last: true};
export const askMoreMessage = {text: MAYA_ASKS_MORE, isBot: true, last: true};

