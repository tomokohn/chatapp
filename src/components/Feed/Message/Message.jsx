import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import botAvatar from '../../../assets/mayas_avatar.png';
import userAvatar from '../../../assets/user_avatar.png';
import TypingLoader from './TypingLoader/TypingLoader';
import {MAYA_TYPING_TIME} from "../../../cosnts";


const applyRadius = (last) => last ? '0 20px 20px 20px' : '20px 20px 20px 0';

const MassageContainer = styled.div`
  display: grid;
  grid-template-columns: 50px auto 50px;
  grid-template-rows: minmax(48px, auto);
  align-items: end;
  justify-content: start;
  direction: ${ props => props.bot ? 'ltr' : 'rtl'};
  margin: ${ props => props.bot ? '0px' : '15px 0'};
`;

const Text = styled.p`
  background: ${ props => props.bot ? '#e9e9e9' : '#4a4a4a'};
  padding: 12px;
  border-radius: ${props => !props.bot ? '20px' : applyRadius(props.last)};
  margin-block-start: 3px;
  margin-block-end: 0;
  grid-column: 2;
  color: ${ props => props.bot ? '#000' : '#fbfbfb'};
  direction: ltr;
  overflow-wrap: break-word;
  word-wrap: break-word;
  max-width: ${ props => `${props.brake - 120}px`}; 
`;

const AvatarImg = styled.img`
  width: 40px;
  padding: 0 5px;
`;

const Message = ({ text, isBot, showAvatar, last }) =>  {
  const [ showTyping, setShowTyping ] = useState(true);
  // constructor() {
  //   super();
  //   this.state = {
  //     showTyping: true
  //   }
  //   this.message = React.createRef();
  // }
  const message = React.createRef();
  const textWidth = null; 
  setTimeout(() => setShowTyping(false), MAYA_TYPING_TIME);
  

  // componentDidMount() {
  //   setTimeout(() => this.setState({showTyping: false,}), MAYA_TYPING_TIME);
  //   this.textWidth =  this.message.current && this.message.current.clientWidth;
  // }

  const avatar = isBot ? botAvatar : userAvatar;
  const imgAltText = isBot ? 'Maya\'s avatar' : 'User\'s avatar';
  return (
    <MassageContainer className="message" bot={isBot} ref={message}>
      {showAvatar && <AvatarImg src={avatar} alt={imgAltText} />}
      {isBot && showTyping ? <TypingLoader last={last} solo={showTyping} /> : <Text last={last} bot={isBot} brake={textWidth}>{text}</Text>}
    </MassageContainer>
  );
};

export default Message;

Message.propTypes = {
  text: PropTypes.string,
  isBot: PropTypes.bool,
  showAvatar: PropTypes.bool,
  last: PropTypes.bool
};