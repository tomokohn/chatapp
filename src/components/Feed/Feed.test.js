import React from 'react';
import renderer from 'react-test-renderer';
import Feed from './Feed';
import Message from "./Message/Message";
import {greetingMessage, askExpressionMessage} from '../../cosnts';


describe('Feed testing', () => {
  test('render Feed', () => {
    const component = renderer.create(<Feed messages={[]} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Feed has 2 messages', () => {
    const messages = [greetingMessage, askExpressionMessage];
    const component = renderer.create(<Feed messages={messages} />);
    const root = component.root;
    expect(root.findAllByType(Message).length).toEqual(2);
  });
});