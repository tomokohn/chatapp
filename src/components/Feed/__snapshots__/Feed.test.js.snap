exports[`Feed testing render Feed 1`] = `
<div
  className="sc-AxgMl jpoiXK feed"
  onScroll={[Function]} />
`;
